### ISSUE TITLE
Demote `Retrying .. after connection broken by ..` log warning to info

### ISSUE THREAD

**Issue by bluetech** [urllib3/urllib3#2583]
### Subject

When `urlopen` retries a request, it issues a warning log:

https://github.com/urllib3/urllib3/blob/37a67739ba9a7835886555744e9dabe3009e538a/src/urllib3/connectionpool.py#L819-L824

I have things set up such that when a warning (or above) is emitted, I get notified.
In my opinion, a request being retried does not warrant a warning, because it is *expected* - network conditions are not ideal, retries are going to happen. If the retry succeeded, great. If it didn't, it will raise an exception which will be handled some way, which *then* may warrant my attention.

This is not hard to workaround, but nonetheless, I think it will be better to change.

I propose:

1. Changing for 2.x (I can send a PR).
2. Backporting to 1.26.x (but can understand if this is not deemed acceptable).

### Environment

```
OS Linux-5.16.14-arch1-1-x86_64-with-glibc2.35
Python 3.9.9
urllib3 1.26.8
```

### Steps to Reproduce

(This is with current urllib3 main):

```py
import logging
import urllib3

logging.basicConfig(level=logging.INFO)

try:
    urllib3.request("GET", "http://example.com:9999", retries=2)
except urllib3.exceptions.MaxRetryError:
    pass
```

### Expected Behavior

The following is printed:

```
INFO:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=0, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fe812e89d20>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /
INFO:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fe812e89ab0>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /
```

### Actual Behavior

The following is printed:

```
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=0, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f49c7831c60>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f49c78319f0>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /
```

**Comment by sigmavirus24** [comment]
Yes, it's true that maybe that shouldn't be a warning but sadly many consumers rely on that.

Namely pip relies on this behavior to give users additional information for debugging without overwhelming them

**Comment by bluetech** [comment]
I kinda figured that such a change is a no-go for 1.26.x, but maybe for 2.x?

Regarding pip, IIUC its default logging level without any verbose/quiet flags is INFO:
https://github.com/pypa/pip/blob/b2b7aefe2f4a40fb34039adaab39897e80df687c/src/pip/_internal/utils/logging.py#L239
so the message will still be displayed, though with INFO instead of WARNING, which seems good to me.
BTW,I found this issue: https://github.com/pypa/pip/issues/8641
It seems that what pip really wants is a way to be notified when a retry happens, and then display its own user-friendly message. But absent that, it relies on the log message.

**Comment by bluetech** [comment]
Correction: reading that pip function further, it seems that for vendored libs like urllib3 they filter at WARNING:
https://github.com/pypa/pip/blob/b2b7aefe2f4a40fb34039adaab39897e80df687c/src/pip/_internal/utils/logging.py#L253-L255
so if we change to INFO it *will* be filtered out.

**Comment by pquentin** [comment]
I'd be happy to change to INFO in 2.x, but would not want to break pip. This is timely as @pradyunsg is trying to improve pip's message, including the one from urllib3: https://github.com/urllib3/urllib3/issues/2580, so maybe there's a way to fill two needs with one deed.

**Comment by pradyunsg** [comment]
:)

FWIW, I'd be fine with changing this to a info log in urllib3 in a 1.x release, as long as the same release also adds some mechanism that pip can use to present those messages. But I don't make the calls on version numbers here.

**Comment by sathieu** [comment]
What about adding a param to the Retry `__init__()`  and passing it in `increment()`?

Proposal: `loglevel=logging.WARNING`.

**Comment by pradyunsg** [comment]
A different approach would be having a callback instead. The default value/implementation for that callback would do the same `logger.warning` call, as is being done currently.

The main point is allowing users to swap it out, which would enable users to do things differently based on their use case. If we want to make it possible to log at the info level, then the callback can be changed to be something that does that.

Now, if you have a callback, you also have a return value out of it. If you want to utilize the callback's return value for something, I think it can be `should_abort_retry = before_retry_callback(...)` whether to abort the retry or not. This would enable doing something like aborting the retry if it's an SSL certificate issue. Having falsy values mean "do the retry" also helps make it easier to implement a function that just returns None, meaning a valid implementation is something that just has a single `logger.warning` statement[^1].

So...

```
before_retry_callback(retries: Retry, err: Exception, url: str) -> Optional[bool]
```

[^1]: Maybe mypy won't be happy? I'm not sure if it enforces that you have a `return None` if you have an `Optional` return value.

**Comment by pradyunsg** [comment]
FWIW, this is also roughly what I'd proposed in #2580 -- so, yea, I think we'd be able to solve both needs in a single shot. :)

**Comment by sigmavirus24** [comment]
> A different approach would be having a callback instead. The default value/implementation for that callback would do the same `logger.warning` call, as is being done currently.

I really dislike that for changing the logging level, etc. Here's why:

I have an application, it has many integrations all of which use urllib3 at some layer of the stack _and_ I have `WARNING` logging enabled for urllib3 purposefully to see Retry errors and build alerting around it. Now, I can't rely on that behaviour because each of the N different libraries have patched that out differently to do different things and I can't rely on that information for alerting because those libraries don't give me the ability to change what they pass in as that callback. Likewise for making it a configurable part of the `Retry` instantiation.

I think a callback works for pip's use-case just fine, and don't object to the idea of a callback to satisfy what pip might want. I don't think the default callback should be `logger.warning` or something that does that because it hurts people using libraries that rely on urllib3.

**Comment by sathieu** [comment]
Proposed backward compatible PR: #2595.

**Comment by jbize** [comment]
So other than the unacceptable connectionpool.log.setLevel(logging.ERROR), what is the workaround?
